helix-core/src/syntax.rs

```diff
@@ -1184,7 +1188,36 @@ struct LocalScope<'a> {
 }
 
 #[derive(Debug)]
-struct HighlightIter<'a> {
+pub struct CharacterHighlightIter<'a>(HighlightIter<'a>);
+
+impl Iterator for CharacterHighlightIter<'_> {
+    type Item = HighlightEvent;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let res = match self.0.next()?.unwrap() {
+            // TODO: use byte slices directly
+            // convert byte offsets to char offset
+            HighlightEvent::Source { start, end } => {
+                let text = self.0.source;
+                let start = text.byte_to_char(ensure_grapheme_boundary_next_byte(text, start));
+                let end = text.byte_to_char(ensure_grapheme_boundary_next_byte(text, end));
+                HighlightEvent::Source { start, end }
+            }
+            event => event,
+        };
+
+        Some(res)
+    }
```

Comment:
> I'm concerned that helix_core is not the appropriate place to be unwrapping. HighlightIter returns Result items so that in the future we can implement cancellation of highlighting (timer based for example). There isn't a concrete plan for that AFAIK so it's hard to say how/where the Result should be handled.

I have moved unwrapping the closure back into `helix-term` in 25427ce altough it make the code simpler if we simply returned `HighlightEvent` from the `HighlightIterator` (panicking directly in the iterator) and then switched to a result once its actually required (because an `unwrap` is required everywhere the `HighlighterIterator` is used all call-sites will have to be changed in the future anyway). What do you think is better

> @archseer does the TODO for using byte slices applies anymore? I think we can just translate bytes to chars in the Iterator implementation for HighlightIter

In `markdown.rs` the highlighting directly indexes into a string and therefore uses byte indices. Using char indices and translating back would be inefficient in that case  so I am not sure there is a better solution right now.
Rewriting the rendering to use byte indices seems to amount to translating byte indices to chars anyways so there might not be a better approach.